lti/tests.py

```diff
@@ -0,0 +1,144 @@
+        Group.objects.create(class_key=1, title="Test title", created_by=self.created_user)
+        group = Group.objects.get(class_key=1)
+        user = User.objects.get(username="user1")
+        lti_initializer.update_roles(user=user, group=group, role="Student")
+        self.assertTrue(user in group.students.all())
+
+    def test_role_is_updated(self):
+        lti_initializer = LtiInitializerView()
+        Group.objects.create(class_key=1, title="Test title", created_by=self.created_user)
+        group = Group.objects.get(class_key=1)
+        user = User.objects.get(username="user1")
+        lti_initializer.update_roles(user=user, group=group, role="Student")
+        self.assertTrue(user in group.students.all())
+        lti_initializer.update_roles(user=user, group=group, role="Teacher")
+        self.assertFalse(user in group.students.all())
+        self.assertTrue(user in group.teachers.all())
+
+    def test_create_lti_user(self):
+        lti_initializer = LtiInitializerView()
+        email = "testMasterFlash@example.com"
+        lti_initializer.create_lti_user(email)
+        user = User.objects.get(email=email)
+        self.assertEqual(user.username, email)
+        self.assertEqual(user.profile.display_name, "testMasterFlash")
+
+    def test_determine_role(self):
+        lti_initializer = LtiInitializerView()
+        teacher_role = "urn:something/something-else/Instructor,urn:something/something-else/Student"
+        self.assertEqual(lti_initializer.determine_role(teacher_role), "Teacher")
+        student_role = "urn:something/something-else/not-Instructor,urn:something/something-else/Student"
+        self.assertEqual(lti_initializer.determine_role(student_role), "Student")
+        # roles of None will also return "Student"
+        self.assertEqual(lti_initializer.determine_role(None), "Student")
+
+    @patch("lti_provider.lti.LTI.verify")
+    def test_dispatch_advises_relaunch(self, mock_validate_request):
+        """ Test missing POST parameters will advise user to relaunch """
+        mock_validate_request.return_value = True
+        response = self.client.post(
+            "/lti/lti_initializer/",
+            {
+                "lis_person_contact_email_primary": "user1@example.com"
+            }
+        )
+        self.assertContains(response, "Your session has expired. Please, relaunch the tool via your canvas course.")
+
+    @patch("lti_provider.lti.LTI.verify")
+    def test_dispatch(self, mock_validate_request):
+        """ Test successful dispatch redirects to home """
+        mock_validate_request.return_value = True
+        response = self.client.post(
+            "/lti/lti_initializer/",
+            {
+                "lis_person_contact_email_primary": "user1@example.com",
+                "custom_canvas_course_id": "327",
+                "context_title": "test title",
+                "ext_roles": "urn:something/something-else/Instructor,urn:something/something-else/Student"
```

Comment:
It might be good to use some realistic values for `ext_roles` to make the test closer to what will actually be sent by the LTI consumer (e.g. Canvas). Here are a few real examples that could be used:

1. As an Instructor in a course (`urn:lti:role:ims/lis/Instructor`):
```
ext_roles=urn:lti:instrole:ims/lis/Instructor,urn:lti:instrole:ims/lis/Student,urn:lti:role:ims/lis/Instructor,urn:lti:sysrole:ims/lis/User
```

2. As a TeachingAssistance in a course (`urn:lti:role:ims/lis/TeachingAssistant`):
```
ext_roles=urn:lti:instrole:ims/lis/Instructor,urn:lti:instrole:ims/lis/Student,urn:lti:role:ims/lis/TeachingAssistant,urn:lti:sysrole:ims/lis/User
```

3. As a Student in a course (`urn:lti:role:ims/lis/Learner`):
```
ext_roles=urn:lti:instrole:ims/lis/Student,urn:lti:role:ims/lis/Learner,urn:lti:sysrole:ims/lis/User
```